aesara/tensor/nnet/basic.py

```diff
@@ -1101,15 +1102,19 @@ def local_logsoftmax_grad(fgraph, node):
             and isinstance(
                 node.inputs[0].owner.inputs[0].owner.op, AdvancedIncSubtensor
             )
+            # the rewrite only applies to legacy SoftmaxGrad
+            and node.op == softmax_grad_legacy
         )
     ):
         # get parameters from unoptimized op
         sm = node.inputs[0].owner.inputs[1]
-        # sm_input = node.inputs[1].owner.inputs[0]
         grads = node.inputs[0].owner.inputs[0]
-        if grads.broadcastable[1] and not sm.broadcastable[1]:
-            grads = aet.alloc(grads, grads.shape[0], sm.shape[1])
-        ret = grads - aet_sum(grads, axis=1, keepdims=True) * sm
+        # TODO: Figure out what the lines below were trying to achieve.
+        #  Non broadcasting across the axis dimension? Why only that one?
+        #  Also no test seems to have been covering these lines
+        # if grads.broadcastable[1] and not sm.broadcastable[1]:
```

Comment:
I couldn't figure out the purpose of the two lines I have commented out. It was introduced in this PR without context and AFAICT no tests cover this situation: https://github.com/Theano/Theano/pull/3363/files

Any ideas?

Reply:
It looks like it might be trying to broadcast `grads` manually so that it matches the `aet_sum(grads, axis=1, keepdims=True) * sm` term.  In doing so, it appears to makes a pretty strong assumption about the dimensionality/axes of the problem.

Reply:
Could this be because of the vector -> row matrix transformation that softmax used to do? Perhaps that's why this line was commented out `# sm_input = node.inputs[1].owner.inputs[0]`